src/sync.c

```diff
@@ -220,23 +222,41 @@ void sync_process(sync_t *st)
         }
         else if (st->cfo_wait == 0)
         {
-            for (i = -2 * PARTITION_WIDTH; i < 2 * PARTITION_WIDTH; ++i)
+            for (cfo = -2 * PARTITION_WIDTH; cfo < 2 * PARTITION_WIDTH; cfo++)
             {
-                int offset2;
-                adjust_ref(st, LB_START + (PM_PARTITIONS * PARTITION_WIDTH) + i, i);
-                offset = find_ref(st, LB_START + (PM_PARTITIONS * PARTITION_WIDTH) + i, 0);
-                if (offset < 0)
-                    continue;
-                // We think we found the start. Check upperband to confirm.
-                adjust_ref(st, UB_END - (PM_PARTITIONS * PARTITION_WIDTH) + i, i);
-                offset2 = find_ref(st, UB_END - (PM_PARTITIONS * PARTITION_WIDTH) + i, 0);
-                if (offset2 == offset)
+                int best_offset = -1;
+                unsigned int best_count = 0;
+
+                memset(offset_count, 0, BLKSZ * sizeof(unsigned int));
+
+                for (i = 0; i <= PM_PARTITIONS; i++)
+                {
+                    adjust_ref(st, cfo + LB_START + i * PARTITION_WIDTH, cfo);
+                    offset = find_ref(st, cfo + LB_START + i * PARTITION_WIDTH, (PM_PARTITIONS-i) & 0x3);
+                    if (offset >= 0)
+                        offset_count[offset]++;
+
+                    adjust_ref(st, cfo + UB_END - i * PARTITION_WIDTH, cfo);
+                    offset = find_ref(st, cfo + UB_END - i * PARTITION_WIDTH, (PM_PARTITIONS-i) & 0x3);
+                    if (offset >= 0)
+                        offset_count[offset]++;
+                }
+
+                for (offset = 0; offset < BLKSZ; offset++)
+                {
+                    if (offset_count[offset] > best_count) {
+                        best_offset = offset;
+                        best_count = offset_count[offset];
+                    }
+                }
+
+                if (best_offset >= 0 && offset_count[best_offset] >= 3)
```

Comment:
Minor comment: you use `offset_count[best_offset]` in here but `best_count` above